enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,231 @@
+
+_Note: We may skip steps depending on the user's existing setup. For example, we may not need to "Verify Kubernetes" on a soft start. The current step/total step numbers will only be approximations_
+
+
+This log would be printed at the correct time later in the code by calling a new function, similar to `out.T`:
+
+```go
+func Step(stepName string, style StyleEnum, format string, a ...V)
+```
+
+In the code itself, this means that this line, which currently looks like this:
+
+```go
+out.T(out.Sparkle, `Using the {{.driver}} driver based on existing profile`, out.V{"driver": ds.String()})
+```
+
+would now be:
+
+```go
+out.Step(out.CreateDriver, out.Sparkle, `Using the {{.driver}} driver based on existing profile`, out.V{"driver": ds.String()})
+```
+
+`out.Step` will be responsible for applying the passed in template, and printing out a JSON encoded version of the step if `--output json` is specified:
+
+```json
+{
+  "Name": "Selecting Driver",
+  "Message": "✨  Using the hyperkit driver based on user configuration\n",
+  "TotalSteps": 9,
+  "CurrentStep": 2,
+  "Type": "Log"
+}
+```
+
+### Stdout - Download Steps
+
+To communicate progress on artifacts as they're being downloaded, we want JSON output that looks something like this during download:
+
+```json
+{
+  "Type": "Download",
+  "Artifact": "preload.tar.gz",
+  "Progress": "10%"
+}
+```
+
+minikube uses the [go-getter](github.com/hashicorp/go-getter) library to download artifacts (preloaded tarballs, ISOs, etc).
+Currently, minikube passes in a [DefaultProgressBar](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/download/download.go#L48) to this library, which is used to communicate download progress to the user.
+
+Instead of passing in `DefaultProgressBar` we should be able to write our own object, something like `JSONOutput`, which will print the current progress of the download in the JSON format specified above instead of showing a progress bar in the terminal.
+
+
+#### Testing Plan
+Both unit tests and integration tests will be required to test these features feature.
+
+Unit tests will cover:
+1. That the JSON output of output steps, both type `Log` and type `Download`, is correct and parsable
+1. That errors are sent to stderr correctly and are parsable
+
+Integration tests will cover:
+1. That in the following cases, if `--output json` is specfied, all logs are correctly in JSON format:
+  * Clean start, with no downloaded artifacts
+  * Soft start
+  * Restart
+   
+
+## Alternatives Considered
+
+### Cloud Events
+
+I briefly looked into using [Cloud Events](https://github.com/cloudevents/spec) to send events to clients, specifically looking at the [Go SDK](https://github.com/cloudevents/sdk-go).
+
+Pros:
+* Standardized way of sending events
+* Supported by CNCF
+
+Cons:
+* Having minikube set up an HTTP server adds extra complexity to this proposal. If clients can instead parse JSON info from stdout/stderr then that would be the simplest solution. 
```

Comment:
AIUI, json over whatever channel you want is also fine. https://github.com/cloudevents/spec/blob/master/spec.md#example even shows one. 

Diffs from your samples to a proper cloud event:
* add an 'id' to each event (so downstream processors can talk about them and deal with dups)
* add a 'source' to each event, like 'https://minikube.sigs.k8s.io/toolEvents' (as opposed to runtime events?)
* add `'specversion': '1.0'`
* change `'Type': 'Download'` to `'type': 'io.k8s.sigs.minikube.download'`, etc
* maybe add 'subject': 'step' (to separate minikube tool's progress steps from future reports that happen at tool runtime)
* optional: add `'time': '<rfc3339 time>'` (not sure if it helps anything here)
* lowercasesmash all your other field names, because that's the cloud events way :)


Example scenario where MK events get routed around:
1. I ask a cloud host to run minikube as part of some CI suite
2. Minikube outputs events on stdout like you're planning
3. The CI runner snarfs minikube's output (and other tools') and routes them back to me over an HTTP server-sent events connection
4. My UI renders those into a rich progress bar

Reply:
Thanks for doing that research @drewpca 

It seems like we could make the JSON struct themselves compatible by adding a few fields, which seems easy enough to do. We'll still print the JSON to stdout/stderr as planned.

cc @balopat WDYT?